### DISCUSSION TITLE
About TempAllocator

### DISCUSSION THREAD

**Discussion by Zylann** [jrouwe/JoltPhysics#969] [Q&A]
I've been looking at how Jolt uses allocators, in particular TempAllocator, as I'm learning how I could use something similar in my own project.

I considered this case:
```cpp
std::vector<int, STLTempAllocator<int>> vec(allocator);
//... later in function...
vec.push_back(1); // allocates some bytes
//... later in function...
vec.push_back(42);  // potentially increases capacity, which allocates more bytes, copies/moves and deallocates old bytes
```
The fact TempAllocator uses a stack means the deallocation during capacity resizing would be illegal, because it's not on top of the stack (instead what's on top is the new capacity-sized buffer).

Am I missing something, or is TempAllocator indeed not supposed to be used in this way and maximum capacity must be known in advance? If so, what kind of strategy would you recommend in situations where max capacity isn't "known"?
What I mean by this, is that the capacity of this vector is still assumed lower than the total size of the TempAllocator. However, it would be unpractical to just do `reserve(total_capacity_of_temp_allocator)`, because that vector could likely not be the only one using the allocator at a given time. It would require some manual management of what capacity to give every container in a given scope/function calls. std::vector also tends to allocate a little more to amortize.
Also, there are other cases in which the container could be `unordered_map` for example, which is a bit harder to estimate.

For some context, an example of area where I use temporary allocs in  my project is mesh generation/processing, in which I often have more than one dynamic container laying around.

**Comment by jrouwe** [comment]
> Am I missing something, or is TempAllocator indeed not supposed to be used in this way and maximum capacity must be known in advance?

Yes, you need to know the max capacity in advance. It needs careful code design to work with this kind of allocator.

> If so, what kind of strategy would you recommend in situations where max capacity isn't "known"?

If you only have 1 container then you can reserve like you say, otherwise you'll need a different allocator strategy. It's a bit out of scope to talk about allocators here, there are plenty of resources on the internet on this subject.
